xrfi chunking

adampbeardsley opened the pull request:
This PR coincides with the opm file chunking branch to enable working in batches for XRFI.

The new wrapper takes in several data files, performs the xrfi algorithm, and stores outputs for the inner parts of the chunk - discarding data from edge files. On the OPM side, the do script will assign overlapping chunks of files to jobs so in the end we get output from all files. 

Because we are discarding the edges, we don't need to pad the data anymore. I've added an option to the algorithm functions to turn off padding, which will save a little time on the median filter.

I've run this through the pipeline on NRAO, but still need to check the outputs. But now's a good time to start getting more eyes on the code.

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -389,6 +389,9 @@ def detrend_medfilt(data, flags=None, Kt=8, Kf=8):
         The box size in frequency (second) dimension to apply medfilt over. Default
         is 8 pixels.
+    pad : bool, optional
+        If True (default), pad the data by mirroring at the edges by the size of
```
Clarify that this is in both dimensions.

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -1639,6 +1658,310 @@ def xrfi_run(ocalfits_file, acalfits_file, model_file, data_file, history,
         uvf.write(outpath, clobber=clobber)
 
 
+def xrfi_h3c_idr2_1_run(ocalfits_files, acalfits_files, model_files, data_files,
```
Again, I'd consider renaming this to something a bit more descriptive.

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -1639,6 +1658,310 @@ def xrfi_run(ocalfits_file, acalfits_file, model_file, data_file, history,
+    None
+
+    """
+    history = 'Flagging command: "' + history + '", Using ' + hera_qm_version_str
```
It's slightly weird that whatever you put in the "history" is labeled as a flagging command, even if that's how it's primarily used. Maybe change the argument name?

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -1639,6 +1658,310 @@ def xrfi_run(ocalfits_file, acalfits_file, model_file, data_file, history,
+    uvf_ag, uvf_agf = xrfi_pipe(uvc_a, alg='detrend_medfilt', Kt=kt_size, Kf=kf_size, xants=xants,
+                                cal_mode='gain', sig_init=sig_init, sig_adj=sig_adj,
+                                label='Abscal gains, round 1.', pad=False,
+                                run_check=run_check,
```
If you define checks = {'run_check': run_check} etc. than you can just ** these kwargs in and save some lines.

jsdillon replied:
Or go broader: make every argument that never changes get passed via a single dict.

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -1639,6 +1658,310 @@ def xrfi_run(ocalfits_file, acalfits_file, model_file, data_file, history,
+                                check_extra=check_extra,
+                                run_check_acceptability=run_check_acceptability)
+
+    # Get the absolute chi-squared values
```
Mislabeled, it's omnical.

jsdillon replied:
Or maybe just confusing

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -1639,6 +1658,310 @@ def xrfi_run(ocalfits_file, acalfits_file, model_file, data_file, history,
+                              check_extra=check_extra,
+                              run_check_acceptability=run_check_acceptability)
+
+    # Get the absolute chi-squared values
```
mislabeled, or confusing at least. It's omnical, not abscal

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -1639,6 +1658,310 @@ def xrfi_run(ocalfits_file, acalfits_file, model_file, data_file, history,
+    end_ind = len(data_files) - ndrop
+    # If we're the first or last job, store all flags for the edge
+    datadir = os.path.dirname(os.path.abspath(data_files[0]))
+    all_files = sorted(glob.glob(os.path.join(datadir, '*sum.uvh5')))
```
Not wild about hardcoding this...

jsdillon replied:
Can we have it figure out the files based on the data files passed in.

adampbeardsley replied:
I don't think so. This is just used to determine the edges of the entire night, which is usually outside the files that are passed in. I've updated the search to be more robust.

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -1639,6 +1658,310 @@ def xrfi_run(ocalfits_file, acalfits_file, model_file, data_file, history,
+    # Determine the actual files to store
+    uvtemp = UVData()
+    uvtemp.read(data_files[0], read_data=False)
+    nintegrations = len(data_files) * uvtemp.Ntimes
```
Can we assume that all files have the same number of integrations?

adampbeardsley replied:
For a given day, yes. I think it would require a correlator restart to have different integrations/file. This logic will probably have to be revisited with BDA... I'm not sure how `Ntimes` is defined there.

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -1639,6 +1658,310 @@ def xrfi_run(ocalfits_file, acalfits_file, model_file, data_file, history,
+    to get better estimate. The metrics and flags from each data product and both
+    rounds are stored in the xrfi_path (which defaults to a subdirectory, see
+    xrfi_path below). Also stored are the a priori flags and combined metrics/flags.
+
```
Explain which input files actually get saved and how the beginning and end of the list are handled.

jsdillon commented on hera_qm/xrfi.py:
```diff
@@ -1639,6 +1658,310 @@ def xrfi_run(ocalfits_file, acalfits_file, model_file, data_file, history,
+                'combined_metrics2.h5': uvf_metrics2, 'combined_flags2.h5': uvf_fws2,
+                'flags2.h5': uvf_combined2}
+
+    # Determine the actual files to store
```
I find this whole procedure a little confusing. You definitely want to comment better what's going on... Would it make more sense for the user to specify which file JD strings are to be completely flagged, RFI flagged, or not written out at all?

adampbeardsley commented:
@jsdillon things got a little disorganized, but I think I addressed all your comments... except the one about function name which I'd like to punt down the road to the restructuring.
Let me know if I missed anything.

adampbeardsley commented:
@jsdillon (or @plaplant ) could I get you to approve again? I was still missing test coverage :(
